### ISSUE TITLE
Custom Output Columns

### ISSUE THREAD

**Issue by amoslai5128** [iisys-hof/map-matching-2#7]
Thank you for giving us this amazing map matching tool, it works so cool.

However, I'm figuring out how to custom the headers & columns in a output result csv, since the current output headers are like:
`id,aborted,duration,track,prepared,match,track_length,prepared_length,match_length`

And, in some cases I need to match the geo points from GPS records with their timestamp in the result, therefore, here is what I want to output:
`id, aborted, duration, ..., *point_timestamp*, *speed*`
_* custom column header_

For your reference, here's the input dataset:
```csv
License, Data Timestamp, Location Timestamp, Lat, Long, Speed Direction, Mileage, ACC Status, Location
Car0001, 2024-03-17 05:44:39, 2024-03-17 05:44:25, 22.34051,114.192995, 0.0, N,29719.2, Y, ABC Street
Car0001, 2024-03-17 05:44:45, 2024-03-17 05:44:25, 22.34051,114.192995, 0.0, N,29719.2, Y, ABC Street
Car3000, 2024-03-17 05:44:39, 2024-03-17 05:44:28, 22.34051,114.192995, 0.0, N,29719.2, Y, EFG Street
Car3000, 2024-03-17 05:44:45, 2024-03-17 05:44:28, 22.34051,114.192995, 23.0, NW12°,29719.3, Y, EFG Street
```

**Comment by addy90** [comment]
Hi @amoslai5128,

thank you for your nice words about my work and your question!

So, I want to understand your issue first. You are trying to match the points 1-by-1 so that for every original point you get the final matched point in the output, without loosing the original columns, right? Technically, this is not possible, because a match, which is a concatenation of road network edges, can contain more or less points than the original track, for example, because curves contain more points and straight lines contain less points than the original track. So how would you output exactly one output point for each input point in such a situation?

Here is an example of what I mean:
![match](https://github.com/user-attachments/assets/13202691-de02-49a1-84b4-7cdc8159c133)
In the left panel is the original track with all points in red. In the middle panel is the track after the trajectory simplification algorithm has been applied in orange. In the right panel is the final match in green.

As you can see, in the curve section in the middle of the track, the match contains much more points than the original track. So for one input point, multiple output points would exist. Furthermore, in the left part of the track on the straight line, the original track contained in one part more points than the match, but the prepared track contained less points.

Let me elaborate on this: Currently, the output (the match) is given as a linestring, and because a 1-to-1 mapping is not feasible, the time and speed information is omitted. It would not make much sense any more in the match, as the points are completely different to the input points, the correlation would not be correct any more.
Furthermore, the trajectory simplification algorithm (which is on by default) already omits the time and speed information, since it is purely geometric. For details, please read section 2.1 in my article: [https://doi.org/10.1111/tgis.13107](https://doi.org/10.1111/tgis.13107).
The matching phase currently also only uses geometric metrics, it also does not use any time information.
The advantage of this approach, to omit any time information, is that any geometric (even hand-drawn) line can be matched with this approach. The disadvantage of omitting the time information is, of course, that it could help to improve the results by incorporating speed metrics. This was a design choice for this approach. We went for more general applicability.

Now for the first good part: The trajectory simplification can be completely disabled by applying `--simplify-track off --median-merge off` (maybe also `--filter-duplicates off` if you have duplicate points). However this slows down the matching performance (both in terms of speed and quality). For details, please refer to section 3.3 in my article: [https://doi.org/10.1111/tgis.13107](https://doi.org/10.1111/tgis.13107).

With the trajectory simplification disabled (or by implementing a trajectory simplification algorithm that does not omit the time information), it would technically be possible to retain the time information throughout the process. However, the first issue with the example image above cannot be solved this way. Outputting the result as a list of points would still not correlate 1-to-1 with the input points and it would be totally questionable where to put which timestamp and speed information. It would be possible to interpolate the values, but this would also be an additional assumption (the match itself already is an assumption, we cannot know if it is correct to the ground truth).

Nevertheless, there exists information about to which point of the match each input point was mapped to. This result can be obtained by applying `--candidates candidates.csv`. It gives a file that contains for each input point the candidates. When the column `is_result` is true, the point is the one from the match. With this file, you can see for each input point each output point 1-to-1. But the complete match linestring still contains more or less points as the input track.

Here is an example of the `is_result = 1` column:
![candidate](https://github.com/user-attachments/assets/7076d4a9-a2fb-449c-9598-6be8a79f6f2a)
In blue there is for each input point of the prepared original track the segment that points to the output point in the match. As you can see, the left input point points exactly on an explicit output point of an edge. The right input point, however, points directly on an edge but without an explicit point on the edge. There was a temporary candidate point that was removed in the output with the `--join-merges` option. This is an additional example why a 1-to-1 point-to-point mapping is not really practical.

When you disable the trajectory simplification completely, you would in fact get for each input point of the original track exactly one output point of the match in the candidates.csv file, although with worse matching quality, as already explained. Moreover, as we have seen, these points are not necessarily contained in the match. But you get for each point a "mapping" to where that point was put in the match. However, keep in mind that the match (in green) generally has a different amount of points as the original track.

With that result, you could theoretically join the results from the candidates.csv file with your input file and append the original timestamp and speed column. Since no points were removed during the matching, you can simply concat the tables (when the input table is sorted by id and timestamp and the candidates table is sorted by id, part_index, set_index, candidate_index).

The result of this can look like this:
![mapping](https://github.com/user-attachments/assets/2f0ad143-2a26-4fac-8a4d-cd85c415f33a)
As you can see, the trajectory simplification was turned off and this way you get a mapping from input to output. Although some of the points are not contained in the match linestring explicitly.

The code that was used to combine the data is here:
```python
# !/usr/bin/env python3
# -*- coding: utf-8 -*-

import numpy as np
import pandas as pd
import geopandas as gpd
import shapely as shp

input_file = "points.csv"
output_file = "merged.csv"
candidates_file = "candidates.csv"

input = pd.read_csv(input_file, parse_dates=["time"])
candidates = pd.read_csv(candidates_file)
results = candidates[candidates["is_result"] == 1].copy()

input.sort_values(by=["id", "time"], inplace=True)
results.sort_values(by=["id", "part_index", "set_index", "candidate_index"], inplace=True)

results.reset_index(drop=True, inplace=True)

assert len(input) == len(results)

projections = gpd.GeoDataFrame(results.drop(columns=["projection"]),
                               geometry=gpd.GeoSeries.from_wkt(results["projection"]))
projections["pointing_coordinate"] = projections['geometry'].apply(
    lambda x: shp.geometry.Point(x.coords[1]) if (len(x.coords) > 1) else np.nan)

input["pointing_coordinate"] = projections["pointing_coordinate"]

input.to_csv(output_file, index=False)
```
You get as output a file where each input point has the output point from the match.
You need the `--simplify-track off --median-merge off --filter-duplicates off--candidates candidates.csv` settings applied.
You might also need to change the parse_dates and sort columns in the code above to the columns you use, e.g., License and Location Timestamp.

With trajectory simplification enabled, this becomes quite impossible, because the median-merge algorithm may introduce new points that were not in the original input track, but this only happens in small noisy situations. However, a 1-to-1 mapping then is practically impossible. You can try to re-map the prepared track points by distance to the original points, but this might not work well when the track goes through the same position multiple times. You can of course also omit these points and make a join with only the points that were also present in the original tracks (the x.coords[0] contains the starting point from the blue segments, which should lie close to the original point), but this needs additional care because of floating point inaccuracies between the original input point and the output point in the candidates.csv.

Please use the new version [v1.0.5](https://github.com/iisys-hof/map-matching-2/releases/tag/v1.0.5) when you want to try out with disabled trajectory simplification. I noticed a bug while preparing the above example and fixed it in advance of this reply.

So TL;DR: We cannot output the original timestamps and speed (or any other columns) from the original track input in the match output because the input and output are too different in amount and position of points. There are, however, ways with certain sacrifices in accuracy and speed, that allow to combine the input and output, as I have shown above. As this is a very specific and individual use-case, I don't really know currently how this could be implemented in this software in a way that it works generally due to all the explanations that I have given. I hope you see it the same way as I do. If not, please let me know!

**Comment by amoslai5128** [comment]
I very appericate for your detailed and patient repsonse and as a new comer to this topic it would take a bit more time for me to study for that. Meanwhile, here's a picture for **better describing to my first question.**

I was trying to find out the data information between the orginal geo-point (red dot) and matched geo-point (green dot), and also you can see we can ingore the dropped geo-point (black bordered red dot).
![Section 1 (1)](https://github.com/user-attachments/assets/8360ae4d-d77c-4c4c-8e67-f8f326cc3d54)

> With the trajectory simplification disabled (or by implementing a trajectory simplification algorithm that does not omit the time information), it would technically be possible to retain the time information throughout the process. However, the first issue with the example image above cannot be solved this way. Outputting the result as a list of points would still not correlate 1-to-1 with the input points and it would be totally questionable where to put which timestamp and speed information. It would be possible to interpolate the values, but this would also be an additional assumption (the match itself already is an assumption, we cannot know if it is correct to the ground truth).

When comes to the actual ouput with simplification enabled, from my understanding (correct me if i'm wrong), is there anyway to do something like this?
![Section 2](https://github.com/user-attachments/assets/eca9e8a0-6b26-4e49-b31d-0e0803b374b6)

Also, how can I find out the route ID for each green geo-points from road network file?

**Comment by addy90** [comment]
Hi, thank you for precising your questions!

I hope that I understand them better now and will try to response to them. First some more comments to your statements:

Concerning your first picture, you get the information from the original geo point to the matched geo point from the `--candidates <filename>` setting, which I already explained in my previous post. You can add `--help` to the application for more information or consult the [help.txt](https://github.com/iisys-hof/map-matching-2/blob/main/help.txt) file.
The candidates file contains in the projection column a segment (as WKT LineString) that points from the original geoposition to the matched geo position. These are the blue arrows from my previous example pictures.

Concerning "dropped original points", there are two ways currently how an original point is "dropped". The first way is via the multiple trajectory simplification algorithms (filter-duplicates, simplify-tracfk, and median-merge). This actually drops points and the median-merge may introduce new ones, as explained in my article in section 2.1: [https://doi.org/10.1111/tgis.13107](https://doi.org/10.1111/tgis.13107).
The second way is the candidate adoption feature, please refer to section 2.3 in my article: [https://doi.org/10.1111/tgis.13107](https://doi.org/10.1111/tgis.13107). This does not "drop" the points from the tack, but collapses multiple points into the same match point, therefore virtually dropping unnecessary points.

See for example this old image from this repository (shows a track from the [Kubicka et. al dataset](https://github.com/iisys-hof/map-matching-2/tree/main/data/kubicka-et-al):
![example-2](https://raw.githubusercontent.com/iisys-hof/map-matching-2/refs/tags/v0.3.0/docs/example-2.png)
You can see that, the red original track was simplified (not shown directly in the image) and from the simplified track the orange projections from original point to match point are shown (they were blue in my previous post, here they are orange). Moreover you can see that some points collapse into the same match points, this is the candidate adoption feature's doing.

So "dropping original points" is done in two ways, one explicit way in the trajectory simplification and one stochastic implicit way in the actual matching optimization algorithm backed by the candidate adoption feature.

You can turn off every of these features individually, not only trajectory simplification, but also candidate adoption! Then no "dopping" occurs at all. However, deactivating both features reduces the matching quality and speed drastically, as can also be seen in my article in section 3.3: [https://doi.org/10.1111/tgis.13107](https://doi.org/10.1111/tgis.13107).

So the `--candidates <filename>` export gives other projections from the original point to the match point (for the rows where `is_result == 1`) depending on your settings.

Concerning your questions:

**Q1**: I think the only valid way would be the one that I explained in my previous post. As soon as you use the trajectory simplification algorithms, the time and speed information is dropped in a way that it cannot be restored correctly anymore. I recommend that you follow my example from my previous post and combine the file that you can generate with the `--candidates <filename>` option (i.e., candidates.csv) with your original data with a script like the one that I posted. Anything you do with "nearest spatial join" would not correctly join, see the picture from above, the projections here are not to the nearest road position but sometimes to a position further away. This map matching software uses a stochastic process to determine the best position (concerning the defined metrics) from a range of positions (called candidates). Please refer to section 1.2 from my article for the basic understanding of what I mean and also to section 2.2 for further understanding of what I mean: [https://doi.org/10.1111/tgis.13107](https://doi.org/10.1111/tgis.13107). So a "nearest spatial join" will probably not work from the matched geo position to the original position. _However_, what I already said in my previous post: You can try to use the projection column from the candidates.csv file; the first coordinate is spatially close to the original position (the difference are mere floating point inaccuracies) for a spatial join with the original data, even with simplified trajectories. A simple nearest spatial join still might not work correctly in all situations, because when you have duplicate points with the same spatial coordinate in the original data, but different timestamps, you don't know with which of these points to join with. And the median-merge algorithm may introduce new points (but within a defined range that you can use as limit in your spatial join, that you can configure yourself, see `--median-merge-distance-tolerance` and disable `--adaptive-median-merge` to be safe. But with a more sophisticated algorithm, or a heuristic (e.g., choose the first point of the duplicates) you might find a solution that is well enough. To be honest, these thoughts give me an idea how a general solution could look like that I could implement in this software. However, since the use-case is very specific, I currently cannot work on this particularly.

**Q2**: You can use the `edge_ids` column that is not on by default with the `--columns <arg>` setting. Please also prepare your network with the `--simplify osm-aware` setting in this case. The second setting makes sure that the ids of the road network edges remain intact during import. The default setting merges edges without retaining the correct ids, because the resulting network is smaller and faster to match on, but when you need the original edge ids, this does not work of course. With the `osm-aware` setting, the edge ids and also tags are retained, which slows down the matching speed slightly, but not the quality. With the extra colum `edge_ids` you get then exported in the matches.csv result file the list of ids of the original edges from the road network that you can then further use. The concatenated geometries then is close to the exported match, although in the beginning and ending, there might be differences, when a match starts and ends within an edge. The whole edge (by id) contains more information than the actual match then. In case you want your matches to fit the edges completely, use the `--export-edges <arg>` option so that the list of edge_ids and the match linestring are of equal geometry information. However, this may lengthen the matches in cases where a track started or ended somewhere in the middle of a road, as then the complete road is returned instead of only the part of the road that was actually driven. Furthermore, there is no "route id" as such, as a route consists of the list of edges (and / or partial edges when `--export-edges` is off, which is the default). With the list of the edge ids, you can, however, join the original route information to the result back. The route information can also be obtained via the `--export-edges-csv-file <arg>` option that you can specify during preparation, then you have a csv file that contains the edge ids and tags that you can merge with the edge ids from the result.

I hope that I could help you further!

Your concepts are very interesting and I see all of your ideas working in practice, although in some cases sacrifices in speed and / or quality are needed. The general use-case of this software is to find a most accurate match for a given track, so use-cases like yours, which are very interesting, however, need more deep diving and some extra work.

Please let me know if I can be of further help. I hope you understand that passing the time / speed information within this software directly to the result is currently out of scope of what I can do, because of the difficulties with the trajectory simplification algorithms.

**Comment by addy90** [comment]
I just wanted to add here that since v1.1.0 there is an additional timestamp output now:
https://github.com/iisys-hof/map-matching-2/blob/e90b2293b1e21a6b745b87e0beab1fabf1587216/help.txt#L637
The start and end timestamp are outputted by default, and there is a new `track_times` and `prepared_times` setting that outputs a list of timestamps for each point of the track and prepared track (after trajectory simplification).
Due to design limitations, currently, the time-zone for the output needs to be specified manually, depending on the input time. See the options following the column option in the help.
Also, the candidates-columns now contain the timestamp for the export respectively.
These enhancements may or may not improve solutions to the original question.
I just wanted to add this information here in case someone reads this.

**Comment by amoslai5128** [comment]
> I just wanted to add here that since v1.1.0 there is an additional timestamp output now:
> 
> [map-matching-2/help.txt](https://github.com/iisys-hof/map-matching-2/blob/e90b2293b1e21a6b745b87e0beab1fabf1587216/help.txt#L637)
> 
> Line 637 in [e90b229](/iisys-hof/map-matching-2/commit/e90b2293b1e21a6b745b87e0beab1fabf1587216)
> 
>  --columns arg (=id,aborted,duration,track,prepared,match,start_time,end_time,track_length,prepared_length,match_length) 
> 
> The start and end timestamp are outputted by default, and there is a new `track_times` and `prepared_times` setting that outputs a list of timestamps for each point of the track and prepared track (after trajectory simplification).
> Due to design limitations, currently, the time-zone for the output needs to be specified manually, depending on the input time. See the options following the column option in the help.
> Also, the candidates-columns now contain the timestamp for the export respectively.
> More information in the release notes of [v1.1.0](https://github.com/iisys-hof/map-matching-2/releases/tag/v1.1.0)
> But please use the latest release: [Latest Release](https://github.com/iisys-hof/map-matching-2/releases/latest)
> These enhancements may or may not improve solutions to the original question.
> I just wanted to add this information here in case someone reads this.

Super! You are the nicest repo owner ever 🥇  and all your efforts are truly appreciated, and I believe it will be incredibly helpful to many people. Please keep up the great work! 👍

**Comment by addy90** [comment]
Hey @amoslai5128, just wanted to let you know that the newest release [v1.2.0](https://github.com/addy90/map-matching-2/releases/tag/v1.2.0) now also has a `match_times` column that interpolates the match times based on the prepared times and the geometric distance (by default), according to the `time-interpolation` setting.
